File: sword_offer/cpp/56_删除链表中重复的结点.cpp
#include <iostream>
#include <vector>
#include <stdio.h>
#include <sys/time.h>

#include "ListNode.h"

/*
struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) :
        val(x), next(NULL) {
    }
};
*/

class Solution {
public:
    // 1. 第一次的方式，利用辅助函数，检测当前节点后续的重复次数
    ListNode* deleteDuplication(ListNode* head) {
        if (head == nullptr) {
            return nullptr;
        }
        
        // 为了检测head是否存在重复，需要新加一个额外的head
        ListNode* new_head = new ListNode(0);
        new_head->next = head;
        ListNode* curr = new_head;
        ListNode* tmp = nullptr;
        
        while (curr != nullptr) {
            // 检测接下来有多少个节点相同，如果不同则返回cnt，并修改tmp指针
            if (checkSame(curr->next, tmp) == 0) {
                curr = curr->next;
            } else {
                curr->next = tmp;
            }
        }
        
        return new_head->next;
    }

    // ===== 工具函数 =====
    // 输出一个结点，返回从这里开始有多少个重复的结点
    // next_head为不重复的下一个结点，这里注意要用引用
    int checkSame(ListNode* head, ListNode* & next_head) {
        int cnt = 0;
        while (head != nullptr && head->next != nullptr) {
            if (head->val != head->next->val) {
                next_head = head->next;
                return cnt;
            } else {
                cnt++;
                head = head->next;
            }
        }
        
        next_head = nullptr;
        return cnt;
    }

    // 测试函数
    void test(std::vector<int>& array) {
        ListNode* result = nullptr;
        struct timeval start, end;
        ListNode* head = ListNode_creatFromVec(array)->next;

        printf("=====\n");
        gettimeofday(&start, nullptr);
        result = this->deleteDuplication(head);
        gettimeofday(&end, nullptr);
        printf("times(us): %ld, result: ", end.tv_usec - start.tv_usec);
        ListNode_print(result, false, true);
        printf("\n");
    }
};

int main(int argc, char* argv[])
{
    std::vector<int> array = {1, 1, 2 ,3, 3, 5, 7, 7, 9};

    Solution s;
    s.test(array);
    
    return 0;
}

File: sword_offer/python/56_删除链表中重复的结点.py
#!/bin/bash/ python3
#-*- coding:utf-8 -*-

import time
import ListNode

# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def deleteDuplication(self, head):
        """
        第一次的方式，利用辅助函数，检测当前节点后续的重复次数
        """
        if head == None:
            return None
        
        new_head = ListNode.ListNode(0)
        new_head.next = head
        curr = new_head
        
        while curr != None:
            cnt, tmp = self.checkSame(curr.next)
            if cnt == 0:
                curr = curr.next
            else:
                curr.next = tmp
                
        return new_head.next

    def checkSame(self, head):
        """
        输入一个结点，输出后边有多个个重复的，以及不重复的那个结点
        """
        cnt = 0
        while head != None and head.next != None:
            if head.val != head.next.val:
                return cnt, head.next
            else:
                cnt += 1
                head = head.next
        
        return cnt, None
    
    def test(self, array):
        """
        测试函数
        """
        head = ListNode.ListNode(0)
        head.creatFromList(array)
        
        print("=====")
        start = time.time()
        result = self.deleteDuplication(head.next)
        end = time.time()
        print("time(us): {:>5.2f}, result: ".format((end - start)*10**6))
        result.listPrint(False)
        

def main():
    array = [1, 1, 2, 3, 3, 5, 7, 7, 9]
    array2 = [1] * 100 + [2] * 200 + [3] + [4] * 400 + [5]

    s = Solution()
    s.test(array)
    s.test(array2)


if __name__ == "__main__":
    main()